lib/DXIL/DxilModule.cpp

```diff
@@ -2081,4 +2111,157 @@ bool DxilModule::IsPrecise(const Instruction *inst) const {
+  }
+  uint64_t rawBefore = shaderFlags.GetShaderFlagsRaw();
+  shaderFlags.CombineShaderFlags(other.shaderFlags);
+  if (rawBefore != shaderFlags.GetShaderFlagsRaw())
+    changed = true;
+  return changed;
+}
+
+void DxilModule::ClearShaderCompatInfo() { m_FuncToShaderCompat.clear(); }
+
+void DxilModule::ComputeShaderCompatInfo() {
+  // Initialize worklist with functions with callers
+  SmallSetVector<llvm::Function *, 8> worklist;
+
+  for (auto &function : GetModule()->getFunctionList()) {
+    if (function.isDeclaration() && !function.isIntrinsic() &&
+        function.getLinkage() ==
+            llvm::GlobalValue::LinkageTypes::ExternalLinkage &&
+        OP::IsDxilOpFunc(&function)) {
+      // update min shader model and shader stage mask per function
+      UpdateFunctionToShaderCompat(&function);
+    } else if (!function.isDeclaration()) {
+      // Initialize worklist with functions with callers.
+      // only used for validator version 1.8+
+      if (DXIL::CompareVersions(m_ValMajor, m_ValMinor, 1, 8) >= 0 &&
+          !function.user_empty())
+        worklist.insert(&function);
+
+      // Skip functions with shader flags already set
+      if (m_FuncToShaderCompat.count(&function) != 0)
+        continue;
+
+      // Collect shader flags for function
+      ShaderCompatInfo &info = m_FuncToShaderCompat[&function];
+      info.shaderFlags = ShaderFlags::CollectShaderFlags(&function, this);
+    }
+  }
+
+  // Propagate ShaderCompatInfo to callers, limit to 1.8+ for compatibility
+  if (DXIL::CompareVersions(m_ValMajor, m_ValMinor, 1, 8) >= 0) {
+    while (!worklist.empty()) {
+      llvm::Function *F = worklist.pop_back_val();
+      ShaderCompatInfo &calleeInfo = m_FuncToShaderCompat[F];
+      // Update callers
+      for (auto U : F->users()) {
+        if (CallInst *CI = dyn_cast<CallInst>(U)) {
+          llvm::Function *caller = CI->getParent()->getParent();
+          // Merge info, if changed and called, add to worklist so we update
+          // any callers of caller as well.
+          if (m_FuncToShaderCompat[caller].Merge(calleeInfo) &&
+              !caller->user_empty())
+            worklist.insert(caller);
+        }
+      }
+    }
+  }
+
+  // We must select the appropriate shader mask for the validator version,
+  // so we don't set any bits the validator doesn't recognize.
+  unsigned ValidShaderMask =
+      (1 << ((unsigned)DXIL::ShaderKind::LastValid + 1)) - 1;
+  if (DXIL::CompareVersions(m_ValMajor, m_ValMinor, 1, 5) < 0) {
+    ValidShaderMask = (1 << ((unsigned)DXIL::ShaderKind::Last_1_4 + 1)) - 1;
+  } else if (DXIL::CompareVersions(m_ValMajor, m_ValMinor, 1, 8) < 0) {
+    ValidShaderMask = (1 << ((unsigned)DXIL::ShaderKind::Last_1_7 + 1)) - 1;
+  } else if (DXIL::CompareVersions(m_ValMajor, m_ValMinor, 1, 8) >= 0) {
+    ValidShaderMask = (1 << ((unsigned)DXIL::ShaderKind::Last_1_8 + 1)) - 1;
+  }
+
+  for (auto &function : GetModule()->getFunctionList()) {
+    if (!function.isDeclaration()) {
```

Comment:
Nit: It would be a bit easier to follow the indentation if this were accomplished using something like ` if (function.isDeclaration()) continue;`

Reply:
ok